src/Init/Lean/PrettyPrinter/Parenthesizer.lean

```diff
@@ -0,0 +1,444 @@
+  -- left-most child, which is the term injected by `trailingNode` in place of the recursion. Left recursion is not an
+  -- issue for the parenthesizer, so we can think of this child being produced by `termParser 0`, or whichever Pratt
+  -- parser is calling us; we only need to know its `mkParen`, which we retrieve from the context. Since the left-most
+  -- child was not actually part of the input to this parser, we should reset the `lbp` after processing it. We also
+  -- need to decrement it by 1 to acommodate the difference between handling of leading and trailing tokens.
+  some lbp ← State.lbp <$> get
+    | panic! "trailingNode.parenthesizer: visited a trailing term without tokens?!";
+  { mkParen := some mkParen, .. } ← read
+    | panic! "trailingNode.parenthesizer called outside of visitParenthesizable call";
+  visitParenthesizable mkParen 0;
+  modify (fun st => { st with lbp := some (lbp - 1) })
+}
+
+@[builtinParenthesizer symbolAux]
+def symbolAux.parenthesizer : Parenthesizer | p =>
+evalOptPrec p.appArg! >>= visitToken
+
+@[builtinParenthesizer symbolNoWsAux] def symbolNoWsAux.parenthesizer := symbolAux.parenthesizer
+@[builtinParenthesizer unicodeSymbol] def unicodeSymbol.parenthesizer := symbolAux.parenthesizer
+
+@[builtinParenthesizer identNoAntiquot]
+def identNoAntiquot.parenthesizer : Parenthesizer | p =>
+visitToken appPrec
+
+@[builtinParenthesizer charLitNoAntiquot] def charLitNoAntiquot.parenthesizer := identNoAntiquot.parenthesizer
+@[builtinParenthesizer strLitNoAntiquot] def strLitNoAntiquot.parenthesizer := identNoAntiquot.parenthesizer
+@[builtinParenthesizer nameLitNoAntiquot] def nameLitNoAntiquot.parenthesizer := identNoAntiquot.parenthesizer
+@[builtinParenthesizer numLitNoAntiquot] def numLitNoAntiquot.parenthesizer := identNoAntiquot.parenthesizer
+@[builtinParenthesizer fieldIdx] def fieldIdx.parenthesizer := identNoAntiquot.parenthesizer
+
+@[builtinParenthesizer many]
+def many.parenthesizer : Parenthesizer | p => do
+stx ← getCur;
+visitArgs $ stx.getArgs.size.forM $ fun _ => visit (p.getArg! 0)
+
+@[builtinParenthesizer many1] def many1.parenthesizer : Parenthesizer | p => do
+stx ← getCur;
+if stx.getKind == nullKind then
+  visitArgs $ stx.getArgs.size.forM $ fun _ => visit (p.getArg! 0)
+else
+  -- can happen with `unboxSingleton = true`
+  visit (p.getArg! 0)
+
+@[builtinParenthesizer Parser.optional]
+def optional.parenthesizer : Parenthesizer | p => do
+visitArgs $ visit (p.getArg! 0)
+
+@[builtinParenthesizer sepBy]
+def sepBy.parenthesizer : Parenthesizer | p => do
+stx ← getCur;
+visitArgs $ (List.range stx.getArgs.size).forM $ fun i => visit (p.getArg! (i % 2))
+
+@[builtinParenthesizer sepBy1] def sepBy1.parenthesizer := sepBy.parenthesizer
+
+@[builtinParenthesizer orelse] def orelse.parenthesizer : Parenthesizer | p => do
+st ← get;
+-- HACK: We have no (immediate) information on which side of the orelse could have produced the current node, so try
```

Comment:
@leodemoura Is it reasonable to assume that `orelse` parameters produce mutually distinct node kinds? Should we try to extract that information here (by normalizing up to `node`) instead of backtracking?

Reply:
Not sure this is true in the current parser. Moreover, we cannot prevent users from doing it when they define their own parsers. BTW, I am happy to add a new exception constructor to `Meta.Exception`. Or we define a new exception type for  `ParenthesizerM`. The `Exception.other "BACKTRACK"` is fine for now, but it may create problems for us in the future (e.g., a different module uses the same trick for performing backtracking).

Reply:
Summary: backtracking looks like a good solution to me. However, we should try to avoid the `Exception.other "BACKTRACK"` hack.